File: yogi-dotnet/test/ConfigurationTest.cs
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

using System;
using System.Runtime.InteropServices;
using Xunit;

namespace test
{
    public class ConfigurationTest : TestCase
    {
        Yogi.Configuration cfg = CreateConfiguration();

        [Fact]
        public void Create()
        {
            MOCK_ConfigurationCreate((ref IntPtr config, int flags) =>
            {
                Assert.Equal((int)Yogi.ConfigurationFlags.None, flags);
                config = pointer;
                return (int)Yogi.ErrorCode.Ok;
            });

            var cfg = new Yogi.Configuration();
            Assert.Equal(Yogi.ConfigurationFlags.None, cfg.Flags);

            MOCK_ConfigurationCreate((ref IntPtr config, int flags) =>
            {
                Assert.Equal((int)Yogi.ConfigurationFlags.MutableCmdLine, flags);
                config = pointer;
                return (int)Yogi.ErrorCode.Ok;
            });

            cfg = new Yogi.Configuration(Yogi.ConfigurationFlags.MutableCmdLine);
        }

        [Fact]
        public void FlagsProperty()
        {
            var cfg = CreateConfiguration(Yogi.ConfigurationFlags.DisableVariables);
            Assert.Equal(Yogi.ConfigurationFlags.DisableVariables, cfg.Flags);
        }

        [Fact]
        public void UpdateFromCommandLine()
        {
            MOCK_ConfigurationUpdateFromCommandLine((IntPtr config, int argc, IntPtr argv, int options) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(2, argc);

                var p = Marshal.ReadIntPtr(argv);
                Assert.NotEqual(IntPtr.Zero, p);
                Assert.Equal("ax", Marshal.PtrToStringAnsi(p));

                p = Marshal.ReadIntPtr(argv + IntPtr.Size);
                Assert.NotEqual(IntPtr.Zero, p);
                Assert.Equal("by", Marshal.PtrToStringAnsi(p));

                Assert.Equal((int)Yogi.CommandLineOptions.BranchGhostMode, options);
                return (int)Yogi.ErrorCode.Ok;
            });

            cfg.UpdateFromCommandLine(new[] { "ax", "by" }, Yogi.CommandLineOptions.BranchGhostMode);
        }

        [Fact]
        public void UpdateFromCommandLineError()
        {
            MOCK_ConfigurationUpdateFromCommandLine((IntPtr config, int argc, IntPtr argv, int options) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                cfg.UpdateFromCommandLine(new[] { "ax", "by", "cz" });
            });
        }

        [Fact]
        public void UpdateFromJson()
        {
            MOCK_ConfigurationUpdateFromJson((IntPtr config, IntPtr json) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal("foo", Marshal.PtrToStringAnsi(json));
                return (int)Yogi.ErrorCode.Ok;
            });

            cfg.UpdateFromJson("foo");
        }

        [Fact]
        public void UpdateFromJsonError()
        {
            MOCK_ConfigurationUpdateFromJson((IntPtr config, IntPtr json) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                cfg.UpdateFromJson("foo");
            });
        }

        [Fact]
        public void UpdateFromFile()
        {
            MOCK_ConfigurationUpdateFromFile((IntPtr config, string filename) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal("foo", filename);
                return (int)Yogi.ErrorCode.Ok;
            });

            cfg.UpdateFromFile("foo");
        }

        [Fact]
        public void UpdateFromFileError()
        {
            MOCK_ConfigurationUpdateFromFile((IntPtr config, string filename) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                cfg.UpdateFromFile("foo");
            });
        }

        [Fact]
        public void Dump()
        {
            MOCK_ConfigurationDump((IntPtr config, ref IntPtr json, ref int jsonsize, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(1, resvars);
                Assert.Equal(-1, indent);
                json = helloBytes;
                jsonsize = helloSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal("hello", cfg.Dump());

            MOCK_ConfigurationDump((IntPtr config, ref IntPtr json, ref int jsonsize, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(0, resvars);
                Assert.Equal(2, indent);
                json = helloBytes;
                jsonsize = helloSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal("hello", cfg.Dump(false, 2));
        }

        [Fact]
        public void DumpError()
        {
            MOCK_ConfigurationDump((IntPtr config, ref IntPtr json, ref int jsonsize, int resvars, int indent) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                cfg.Dump();
            });
        }

        [Fact]
        public void ToStringMethod()
        {
            MOCK_ConfigurationDump((IntPtr config, ref IntPtr json, ref int jsonsize, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(1, resvars);
                Assert.Equal(-1, indent);
                json = helloBytes;
                jsonsize = helloSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal("hello", cfg.ToString());
        }

        [Fact]
        public void ToJson()
        {
            MOCK_ConfigurationDump((IntPtr config, ref IntPtr json, ref int jsonsize, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(1, resvars);
                Assert.Equal(-1, indent);
                json = validJsonBytes;
                jsonsize = validJsonSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(5, cfg.ToJson()["x"]);

            MOCK_ConfigurationDump((IntPtr config, ref IntPtr json, ref int jsonsize, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(0, resvars);
                Assert.Equal(-1, indent);
                json = validJsonBytes;
                jsonsize = validJsonSize;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(5, cfg.ToJson(false)["x"]);
        }

        [Fact]
        public void WriteToFile()
        {
            MOCK_ConfigurationWriteToFile((IntPtr config, string filename, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(1, resvars);
                Assert.Equal(-1, indent);
                Assert.Equal("foo", filename);
                return (int)Yogi.ErrorCode.Ok;
            });

            cfg.WriteToFile("foo");

            MOCK_ConfigurationWriteToFile((IntPtr config, string filename, int resvars, int indent) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal(0, resvars);
                Assert.Equal(2, indent);
                Assert.Equal("foo", filename);
                return (int)Yogi.ErrorCode.Ok;
            });

            cfg.WriteToFile("foo", false, 2);
        }

        [Fact]
        public void WriteToFileError()
        {
            MOCK_ConfigurationWriteToFile((IntPtr config, string filename, int resvars, int indent) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                cfg.WriteToFile("foo");
            });
        }

        [Fact]
        public void Validate()
        {
            MOCK_ConfigurationValidate((IntPtr config, string section, IntPtr schema) =>
            {
                Assert.Equal(pointer, config);
                Assert.Equal("foo", section);
                Assert.Equal(schema, config);
                return (int)Yogi.ErrorCode.Ok;
            });

            var scm = CreateConfiguration();
            cfg.Validate("foo", scm);
        }

        [Fact]
        public void ValidateError()
        {
            MOCK_ConfigurationValidate((IntPtr config, string section, IntPtr schema) =>
            {
                Assert.Equal(pointer, config);
                Assert.Null(section);
                Assert.Equal(schema, config);
                return (int)Yogi.ErrorCode.Unknown;
            });

            var scm = CreateConfiguration();

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                cfg.Validate(scm);
            });
        }
    }
}

File: yogi-cpp/test/yogi/configuration_test.cc
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <test/common.h>

#include <yogi_core.h>

class ConfigurationTest : public Test {};

TEST_F(ConfigurationTest, Create) {
  MOCK_ConfigurationCreate([](void** config, int flags) {
    EXPECT_NE(config, nullptr);
    EXPECT_EQ(flags, YOGI_CFG_NONE);
    return YOGI_OK;
  });

  EXPECT_TRUE(!!Configuration::create());

  MOCK_ConfigurationCreate([](void**, int flags) {
    EXPECT_EQ(flags, YOGI_CFG_MUTABLE_CMD_LINE);
    return YOGI_OK;
  });

  EXPECT_TRUE(!!Configuration::create(yogi::ConfigurationFlags::kMutableCmdLine));
}

TEST_F(ConfigurationTest, CreateError) {
  MOCK_ConfigurationCreate([](void**, int flags) {
    EXPECT_EQ(flags, YOGI_CFG_MUTABLE_CMD_LINE);
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(Configuration::create(ConfigurationFlags::kMutableCmdLine), FailureException);
}

TEST_F(ConfigurationTest, CreateFromJson) {
  MOCK_ConfigurationCreate([](void** config, int flags) {
    EXPECT_NE(config, nullptr);
    EXPECT_EQ(flags, YOGI_CFG_NONE);
    *config = kPointer;
    return YOGI_OK;
  });

  MOCK_ConfigurationUpdateFromJson([](void* config, const char* json) {
    EXPECT_EQ(config, kPointer);
    EXPECT_STREQ(json, "foo");
    return YOGI_OK;
  });

  EXPECT_TRUE(!!yogi::Configuration::create("foo"));
}

TEST_F(ConfigurationTest, Flags) {
  auto cfg = create_configuration(ConfigurationFlags::kMutableCmdLine);
  EXPECT_EQ(cfg->flags(), ConfigurationFlags::kMutableCmdLine);
}

TEST_F(ConfigurationTest, UpdateFromCommandLine) {
  auto cfg = create_configuration();

  MOCK_ConfigurationUpdateFromCommandLine([](void* config, int argc, const char* const* argv, int options) {
    EXPECT_EQ(config, kPointer);
    EXPECT_EQ(argc, 5);
    EXPECT_NE(argv, nullptr);
    EXPECT_EQ(options, YOGI_CLO_BRANCH_GHOST_MODE);
    return YOGI_OK;
  });

  const char* const argv[] = {"foo"};
  cfg->update_from_command_line(5, argv, CommandLineOptions::kBranchGhostMode);
}

TEST_F(ConfigurationTest, UpdateFromCommandLineError) {
  auto cfg = create_configuration();
  MOCK_ConfigurationUpdateFromCommandLine([](void*, int, const char* const*, int) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(cfg->update_from_command_line(5, nullptr), FailureException);
}

TEST_F(ConfigurationTest, UpdateFromJson) {
  auto cfg = create_configuration();

  MOCK_ConfigurationUpdateFromJson([](void* config, const char* json) {
    EXPECT_EQ(config, kPointer);
    EXPECT_STREQ(json, "foo");
    return YOGI_OK;
  });

  cfg->update_from_json("foo");
}

TEST_F(ConfigurationTest, UpdateFromJsonError) {
  auto cfg = create_configuration();
  MOCK_ConfigurationUpdateFromJson([](void*, const char*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(cfg->update_from_json("foo"), FailureException);
}

TEST_F(ConfigurationTest, UpdateFromFile) {
  auto cfg = create_configuration();

  MOCK_ConfigurationUpdateFromFile([](void* config, const char* filename) {
    EXPECT_EQ(config, kPointer);
    EXPECT_STREQ(filename, "foo");
    return YOGI_OK;
  });

  cfg->update_from_file("foo");
}

TEST_F(ConfigurationTest, UpdateFromFileError) {
  auto cfg = create_configuration();
  MOCK_ConfigurationUpdateFromFile([](void*, const char*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(cfg->update_from_file("foo"), FailureException);
}

TEST_F(ConfigurationTest, Dump) {
  auto cfg = create_configuration();

  MOCK_ConfigurationDump([](void* config, const char** json, int* jsonsize, int resvars, int indent) {
    EXPECT_EQ(config, kPointer);
    EXPECT_NE(nullptr, json);
    EXPECT_NE(nullptr, jsonsize);
    EXPECT_EQ(resvars, YOGI_TRUE);
    EXPECT_EQ(indent, -1);
    *json     = "foo";
    *jsonsize = static_cast<int>(strlen(*json) + 1);
    return YOGI_OK;
  });
  EXPECT_EQ(cfg->dump(), "foo");

  MOCK_ConfigurationDump([](void*, const char** json, int* jsonsize, int resvars, int indent) {
    EXPECT_EQ(resvars, YOGI_FALSE);
    EXPECT_EQ(indent, 5);
    *json     = "foo";
    *jsonsize = static_cast<int>(strlen(*json) + 1);
    return YOGI_OK;
  });
  EXPECT_EQ(cfg->dump(false, 5), "foo");

  MOCK_ConfigurationDump([](void*, const char** json, int* jsonsize, int resvars, int indent) {
    EXPECT_EQ(resvars, YOGI_FALSE);
    EXPECT_EQ(indent, -1);
    *json     = "foo";
    *jsonsize = static_cast<int>(strlen(*json) + 1);
    return YOGI_OK;
  });
  EXPECT_EQ(cfg->dump(false), "foo");

  MOCK_ConfigurationDump([](void*, const char** json, int* jsonsize, int resvars, int indent) {
    EXPECT_EQ(resvars, YOGI_TRUE);
    EXPECT_EQ(indent, 5);
    *json     = "foo";
    *jsonsize = static_cast<int>(strlen(*json) + 1);
    return YOGI_OK;
  });
  EXPECT_EQ(cfg->dump(5), "foo");

  cfg = create_configuration(ConfigurationFlags::kDisableVariables);
  MOCK_ConfigurationDump([](void*, const char** json, int* jsonsize, int resvars, int) {
    EXPECT_EQ(resvars, YOGI_FALSE);
    *json     = "foo";
    *jsonsize = static_cast<int>(strlen(*json) + 1);
    return YOGI_OK;
  });
  EXPECT_EQ(cfg->dump(), "foo");
}

TEST_F(ConfigurationTest, DumpError) {
  auto cfg = create_configuration();

  MOCK_ConfigurationDump([](void*, const char**, int*, int, int) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(cfg->dump(), FailureException);

  MOCK_ConfigurationDump([](void*, const char**, int*, int, int) { return YOGI_OK; });
  EXPECT_THROW(cfg->dump(-1), FailureException);
}

TEST_F(ConfigurationTest, WriteToFile) {
  auto cfg = create_configuration();

  MOCK_ConfigurationWriteToFile([](void* config, const char* filename, int resvars, int indent) {
    EXPECT_EQ(config, kPointer);
    EXPECT_STREQ(filename, "foo");
    EXPECT_EQ(resvars, YOGI_TRUE);
    EXPECT_EQ(indent, -1);
    return YOGI_OK;
  });
  cfg->write_to_file("foo");

  MOCK_ConfigurationWriteToFile([](void*, const char*, int resvars, int indent) {
    EXPECT_EQ(resvars, YOGI_FALSE);
    EXPECT_EQ(indent, 5);
    return YOGI_OK;
  });
  cfg->write_to_file("foo", false, 5);

  MOCK_ConfigurationWriteToFile([](void*, const char*, int resvars, int indent) {
    EXPECT_EQ(resvars, YOGI_FALSE);
    EXPECT_EQ(indent, -1);
    return YOGI_OK;
  });
  cfg->write_to_file("foo", false);

  MOCK_ConfigurationWriteToFile([](void*, const char*, int resvars, int indent) {
    EXPECT_EQ(resvars, YOGI_TRUE);
    EXPECT_EQ(indent, 5);
    return YOGI_OK;
  });
  cfg->write_to_file("foo", 5);

  cfg = create_configuration(ConfigurationFlags::kDisableVariables);
  MOCK_ConfigurationWriteToFile([](void*, const char*, int resvars, int) {
    EXPECT_EQ(resvars, YOGI_FALSE);
    return YOGI_OK;
  });
  cfg->write_to_file("foo");
}

TEST_F(ConfigurationTest, WriteToFileError) {
  auto cfg = create_configuration();

  MOCK_ConfigurationWriteToFile([](void*, const char*, int, int) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(cfg->write_to_file("foo"), FailureException);

  MOCK_ConfigurationWriteToFile([](void*, const char*, int, int) { return YOGI_OK; });
  EXPECT_THROW(cfg->write_to_file("foo", -1), FailureException);
}

TEST_F(ConfigurationTest, Validate) {
  MOCK_ConfigurationCreate([](void** cfg, int) {
    *cfg = reinterpret_cast<void*>(0x1234);
    return YOGI_OK;
  });
  auto cfg = Configuration::create();

  MOCK_ConfigurationCreate([](void** cfg, int) {
    *cfg = reinterpret_cast<void*>(0x5678);
    return YOGI_OK;
  });
  auto scm = Configuration::create();

  MOCK_ConfigurationValidate([](void* config, const char* section, void* schema) {
    EXPECT_EQ(config, reinterpret_cast<void*>(0x1234));
    EXPECT_STREQ(section, "foo");
    EXPECT_EQ(schema, reinterpret_cast<void*>(0x5678));
    return YOGI_OK;
  });
  cfg->validate("foo", *scm);

  MOCK_ConfigurationValidate([](void*, const char* section, void*) {
    EXPECT_EQ(section, nullptr);
    return YOGI_OK;
  });
  cfg->validate(*scm);
}

TEST_F(ConfigurationTest, ValidateError) {
  auto cfg = create_configuration();
  auto scm = create_configuration();

  MOCK_ConfigurationValidate([](void*, const char*, void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(cfg->validate(*scm), FailureException);
}
